### PR TITLE
[mlir][gpu] Separate the barrier elimination code from transform ops

### PR DESCRIPTION
**Pull request by spaceotter**
Allows the barrier elimination code to be run from C++ as well. The code
from transforms dialect is copied as-is, the pass and populate functions
have beed added at the end.

### CODE REVIEW

**Comment by Groverkss** [comment]
The changes moving the file look mechanical. Could you mention what changes are simply moving them to a new file and what changes need to be reviewed?

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td [hunk 1]
```diff
@@ -37,6 +37,20 @@ def GpuMapParallelLoopsPass
   let dependentDialects = ["mlir::gpu::GPUDialect"];
 }
 
+def GpuEliminateBarriers
+    : Pass<"gpu-eliminate-barriers", "mlir::func::FuncOp"> {
+  let summary = "Erase unecessary barriers";
```
Comment (JackWolfard) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:42]: nit: sp unecessary -> unnecessary

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td [hunk 2]
```diff
@@ -37,6 +37,20 @@ def GpuMapParallelLoopsPass
...
+  let constructor = "mlir::createGpuEliminateBarriersPass()";
+  let description = [{
+    Erase barriers that are unneeded because there are no dependent side effects
+    across the barrier.
+  }];
+  let dependentDialects = [
+    "mlir::gpu::GPUDialect", "mlir::scf::SCFDialect",
```
Comment (antiagainst) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:49]: Keep this alphabetically sorted?

**Review by antiagainst — APPROVED** [review]
I assume this is just shuffling code around; so didn't look at the EliminateBarriers.cpp

File: mlir/lib/Dialect/GPU/Transforms/EliminateBarriers.cpp [hunk 3]
```diff
@@ -0,0 +1,614 @@
+#include "mlir/Dialect/Func/IR/FuncOps.h"
```
Comment (qedawkins) [inline mlir/lib/Dialect/GPU/Transforms/EliminateBarriers.cpp:1]: This file needs a license header
```
//===- EliminateBarriers.cpp - Eliminate extra barriers --===//
//
// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
// See https://llvm.org/LICENSE.txt for license information.
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
//
//===----------------------------------------------------------------------===//
//
// <Detailed description>
//
//===----------------------------------------------------------------------===//

```

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.h [hunk 4]
```diff
@@ -165,6 +165,10 @@ void populateGpuDecomposeMemrefsPatterns(RewritePatternSet &patterns);
 /// Pass decomposes memref ops inside `gpu.launch` body.
 std::unique_ptr<Pass> createGpuDecomposeMemrefsPass();
 
+void populateGpuEliminateBarriersPatterns(RewritePatternSet &patterns);
+
+std::unique_ptr<OperationPass<func::FuncOp>> createGpuEliminateBarriersPass();
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.h:170]: Please add documentation to all top-level symbols.

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td (hunk 2 above)
Comment (ftynse) [reply mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:49]: And one per line.

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td [hunk 5]
```diff
@@ -37,6 +37,20 @@ def GpuMapParallelLoopsPass
   let dependentDialects = ["mlir::gpu::GPUDialect"];
 }
 
+def GpuEliminateBarriers
+    : Pass<"gpu-eliminate-barriers", "mlir::func::FuncOp"> {
+  let summary = "Erase unecessary barriers";
+  let constructor = "mlir::createGpuEliminateBarriersPass()";
+  let description = [{
+    Erase barriers that are unneeded because there are no dependent side effects
+    across the barrier.
```
Comment (ftynse) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:46]: Carrying over the documentation from the pattern may help the reader understand how this is implemented.

```
If a barrier does not enforce any conflicting
/// pair of memory effects, including a pair that is enforced by another
/// barrier, it is unnecessary and can be removed. Adapted from
/// "High-Performance GPU-to-CPU Transpilation and Optimization via High-Level
/// Parallel Constructs" by Moses, Ivanov, Domke, Endo, Doerfert, and Zinenko in
/// PPoPP 2023 and implementation in Polygeist.
```

**Review by ftynse — APPROVED** [review]
LGTM with comments addressed.

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.h (hunk 4 above)
Comment (joker-eph) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.h:170]: The create method should be auto generated (remove from the .td)

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td [hunk 6]
```diff
@@ -37,6 +37,20 @@ def GpuMapParallelLoopsPass
   let dependentDialects = ["mlir::gpu::GPUDialect"];
 }
 
+def GpuEliminateBarriers
+    : Pass<"gpu-eliminate-barriers", "mlir::func::FuncOp"> {
+  let summary = "Erase unecessary barriers";
+  let constructor = "mlir::createGpuEliminateBarriersPass()";
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:43]: Removing this line will make this autogenerated

File: mlir/include/mlir/Dialect/GPU/Transforms/Passes.td [hunk 7]
```diff
@@ -37,6 +37,20 @@ def GpuMapParallelLoopsPass
...
+  let description = [{
+    Erase barriers that are unneeded because there are no dependent side effects
+    across the barrier.
+  }];
+  let dependentDialects = [
+    "mlir::gpu::GPUDialect", "mlir::scf::SCFDialect",
+    "mlir::memref::MemRefDialect"
```
Comment (joker-eph) [inline mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:50]: Why do we need any of these?

Comment (spaceotter) [reply mlir/include/mlir/Dialect/GPU/Transforms/Passes.td:50]: They are used in the EliminateBarriers.cpp file.

**Comment by joker-eph** [comment]
This broke the build:
https://lab.llvm.org/buildbot/#/builders/61/builds/51692/steps/5/logs/stdio
(You should have got an email?)

**Comment by spaceotter** [comment]
I did get an email. Don't know why this didn't appear before, my local build never encountered this.

**Comment by joker-eph** [comment]
Likely -DBUILD_SHARED_LIBS=ON
This indicates missing dependencies in cmake in general.
(On my phone, can’t help more just now, if you don’t figure out, please revert and I can look later)

**Comment by spaceotter** [comment]
This might be all that is needed https://github.com/llvm/llvm-project/pull/72019

**Pull request merged**